keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,280 @@
+---
+title: Object Bucket Provisioning
+authors:
+  - "@jeffvance"
+  - "@copejon"
+owning-sig: "sig-storage"
+reviewers:
+  - "@saad-ali"
+  - "@alarge"
+  - "@erinboyd"
+  - "@guymguym"
+  - "@travisn"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-25
+last-updated: 2020-02-27
+status: provisional
+---
+
+# Object Bucket Provisioning
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Vocabulary](#vocabulary)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+      - [Admin](#admin)
+      - [User](#user)
+  - [System Configuration](#system-configuration)
+  - [Workflows](#workflows)
+    - [Create Bucket](#create-bucket)
+    - [Delete Bucket](#delete-bucket)
+    - [Grant Bucket Access for Static Brownfield](#grant-bucket-access-for-static-brownfield)
+    - [Grant Bucket Access for Dynamic Brownfield](#grant-bucket-access-for-dynamic-brownfield)
+    - [Revoke Bucket Access (needs to handle stratic and dynamic brownfield!)](#revoke-bucket-access-needs-to-handle-stratic-and-dynamic-brownfield)
+  - [Custom Resource Definitions](#custom-resource-definitions)
+      - [ObjectBucketClaim (OBC)](#objectbucketclaim-obc)
+      - [ObjectBucket (OB)](#objectbucket-ob)
+      - [BucketClass](#bucketclass)
+<!-- /toc -->
+
+# Summary
+
+This proposal introduces the Container Object Storage Interface (COSI), whose purpose is to provide a familiar and standardized method of provisioning object storage buckets in an manner agnostic to the storage vendor. The COSI environment is comprised of Kubernetes CRDs, operators to manage these CRDs, and a gRPC interface by which these operators communicate with vendor drivers.  This design is of course heavily inspired by the Kubernetes’ implementation of the Container Storage Interface (CSI).  However, bucket management lacks some of the notable requirements of block and file provisioning and allows for an architecture with lower overall complexity.  This proposal does not include a standardized protocol or abstraction of storage vendor APIs.  
+
+## Motivation 
+
+File and block are first class citizens within the Kubernetes ecosystem.  Object, though different on a fundamental level and lacking an open, committee controlled interface like POSIX, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to elevate buckets to a community supported feature.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage.
+
+## Goals
++ Define a control plane API in order to standardize and formalize a Kubernetes bucket provisioning
++ Minimize privileges required to run a storage driver.
++ Minimize technical ramp-up for storage vendors.
```

Comment:
Unlike CSI which has potentially hundreds of vendor-drivers, how many object-store drivers will really exist?  I just want to make sure we don't over-complicate the design "because CSI" when the goals may actually be pretty different.

Reply:
Way more than I thought.
In no particular order:
**vendors**: amazon, google, red hat (separate from ibm), ibm, emc, netapp, minio, microsoft, dell (separate from emc), swiftstack, cloudian, caringo, data direct networks, HGST, hitachi, huawei, pure, scality, western digital, qumulo, oracle, rackspace, digitalocean, alibaba, vultr, zadara, exablox, tarmin, gScale, oneblox. openIO, quantastor, igneous, basho, etergo, hyve, newisys, supermicro, seagate, hpe, 

**tech**: ceph rgw, aws s3, azure blobs, leofs, swift, minio, edgefs

There may be some dups. I suspect a lot of these are at least loosely based on s3, so an `s3-compatible` protocol would apply to many.

Reply:
How many of those vendors will use the same driver (when they have same tech) vs have totally distinct drivers?

Reply:
I hadn't considered shared drivers but that's an interesting thought.
Although the current proposal states all drivers live in the _cosi-system_ ns, that is up for debate as there are opinions that drivers (and their sidecars) should be namespace isolated (likely in the same ns as the object store when it resides in the k8s cluster).
If this KEP is accepted sig-storage would likely support the COSI sidecar controller and my thought was that each vendor would write their own driver. Of course many of the `s3-compatible` drivers would be similar and could technically be shared. I supposed sig-storage could even support an `s3-compatible` driver and let vendors who want to differentiate themselves write their own?

Reply:
While I expect that bucket creation and deletion might fold into a few uses cases, I'm not clear that would be the case for IAM-style parameters.  They way that identify is communicated to the bucket storage system is likely mostly standardized, but not how you actually setup  those identifies and define their access control.